zerver/tests/test_push_notifications.py

```diff
@@ -992,6 +995,73 @@ def test_send_remove_notifications_to_bouncer(self) -> None:
+            message=message
+        )
+
+        non_encrypted_android_device = PushDeviceToken.objects.create(
+            kind=PushDeviceToken.GCM,
+            token=hex_to_b64(u'eeee'),
+            user=self.user_profile
+        )
+
+        encrypted_android_device = PushDeviceToken.objects.create(
+            kind=PushDeviceToken.GCM,
+            token=hex_to_b64(u'dddd'),
+            notification_encryption_key=generate_encryption_key(),
+            user=self.user_profile
+        )
+
+        non_encrypted_apple_device = PushDeviceToken.objects.create(
+            kind=PushDeviceToken.APNS,
+            token=hex_to_b64(u'aaaa'),
+            user=self.user_profile
+        )
+
+        encrypted_apple_device = PushDeviceToken.objects.create(
+            kind=PushDeviceToken.APNS,
+            token=hex_to_b64(u'bbbb'),
+            notification_encryption_key=generate_encryption_key(),
+            user=self.user_profile
+        )
+
+        missed_message = {
+            'message_id': message.id,
+            'trigger': 'private_message',
+        }
+        with mock.patch('zerver.lib.push_notifications.get_message_payload_apns',
+                        return_value={'apns': True}), \
+                mock.patch('zerver.lib.push_notifications.get_message_payload_gcm',
+                           return_value=({'gcm': True}, {})), \
+                mock.patch('zerver.lib.push_notifications.encrypt_apns_payload',
+                           return_value={'en_apns': True}), \
+                mock.patch('zerver.lib.push_notifications.encrypt_gcm_payload',
+                           return_value={'en_gcm': True}), \
+                mock.patch('zerver.lib.push_notifications'
+                           '.send_apple_push_notification') as mock_send_apple, \
+                mock.patch('zerver.lib.push_notifications'
+                           '.send_android_push_notification') as mock_send_android, \
+                self.settings(PUSH_NOTIFICATION_ENCRYPTION=True):
+
+            handle_push_notification(self.user_profile.id, missed_message)
+            mock_send_apple.assert_any_call(self.user_profile.id,
+                                            [encrypted_apple_device],
+                                            {'en_apns': True},
+                                            remote=False)
+            mock_send_apple.assert_any_call(self.user_profile.id,
+                                            [non_encrypted_apple_device],
+                                            {'apns': True})
+            self.assertEqual(mock_send_apple.call_count, 2)
+            mock_send_android.assert_called_with([encrypted_android_device],
+                                                 {'en_gcm': True}, {}, False)
+            mock_send_android.assert_any_call([non_encrypted_android_device],
+                                              {'gcm': True}, {})
+            self.assertEqual(mock_send_android.call_count, 2)
```

Comment:
It'd be really nice if these tests verified the format of the encrypted payloads being sent to APNS / GCM; specifically, I'd like the test to really clearly verify exactly what fields are sent outside the encrypted blob.

Reply:
Also, we should extend this test to take the verified encrypted payload sent via the push notifications bouncer, captures what the device would receive, and decrypts that using a function we write that is basically intended as sample code for what we'll ask the mobile app developers to write, and returns the exact payload that was sent to the unencrypted endpoint.

(Basically confirming that one can reverse the encryption transformation that we apply on the client side).

One detail that I'm not sure how we should be handling is the question of how the device-side code will know which Zulip realm (and thus encryption key) sent the notification if it is registered for notifications with multiple realms (and thus, critically, which decryption key to use to decrypt the notice).  We may need to extend the protocol to handle that; but before we do so we should write this test in a way that supports that idea.